### PR TITLE
feat: Implement a two-phase init which allows us to write normal init scripts

### PR DESCRIPTION
**Pull request by chipbuster**
#### Description

Implement a two-phase init procedure in starship. The first phase causes the shell to source a subshell, while the second phase (in the subshell) prints the main init script.

This allows us to have nice init scripts with good styling, comments, and no pile of semicolons. Even better, it works as a drop-in replacement, so we don't need to update the docs.

#### Motivation and Context
One of the problems with using `eval` directly is that our scripts all end up on one line. This leads to several problems when it comes to script development:

 - You cannot comment the script, because `#` characters continue to the end of the line, which is the rest of the script when it's all on one line.
 - Debugging is a nightmare--all you know is that there's an error *somewhere* within the script
 - You're forced to place semicolons in places which are not intuitive, even for experienced script writers (you have to have semicolons after every conditional statement, but if you put it after a `then`, it breaks badly and I don't know why)

This can be solved by using subshells to capture output from `source`. However, this leads to an additional problem: if we change our command to `source <(starship init $0)`, the users will have to change their shell rcfiles, which is going to be a bit of a hassle.

Instead, this update allows us to get the benefits of sourcing configs without any of the downside. It does this with a two-phase algorithm (the code is demoed with zsh syntax, but the same idea holds for all shells):

**Main invocation**

The output of `starship init zsh` is `source <(starship init zsh --print-full-init)`. This causes zsh to run a subshell which runs `starship init zsh --print-full-init`, then feed that output to `source` as a file.

**Inner invocation**

The output of `starship init zsh --print-full-init` is the zsh init script (`ZSH_INIT` in the code), complete with whitespace and newlines.

Since this is passed as a file to `source`, whitespace is retained and we can do things like comments.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Screenshots (if appropriate):

Ideally, nothing should change. However, error reporting and recovery is a lot better.

In the following screenshot, I put the same malformed zsh statement into the init script. On the right is what the error looks like before this PR, on the left is after.

image

Note that the left side (the "after" image) tells us the error is on line 20, which we can use to debug. Meanwhile, the right side (the "before") just tells us that something has gone wrong at one of the many conditional tokens in the init script.

#### How Has This Been Tested?

- [x] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

So one major, major caveat here: this is definitely pushing a little into dark shell sorcery, and I'm not 100% confident that this solution is portable across all versions of all shells. I already hit one weird edge case due to the fact that MacOS's version of `bash` is ancient, and there may be more in an older version of zsh, or maybe even in fish.

An alternative is to simply use `source <(starship init <shell>)`, which I am fairly confident will almost always work, but requires users to change their rcfiles. On one hand, we're pre-1.0, so a little breakage is to be expected, but on the other, I don't like springing surprises on people.

#### Checklist:

- [ ] I have updated the documentation accordingly. (not needed)
- [ ] I have updated the tests accordingly.

### CODE REVIEW

**Comment by chipbuster** [comment]
A system like this would have allowed #154 to be merged without breaking the init.

**Comment by chipbuster** [comment]
Currently waiting on #183 to merge, then I'll rebase and un-draft-mode this thing.

File: src/init.rs [hunk 1]
```diff
@@ -1,26 +1,39 @@
...
+pub fn init_stub(shell_name: &str) {
     log::debug!("Shell name: {}", shell_name);
 
     let shell_basename = Path::new(shell_name).file_stem().and_then(OsStr::to_str);
 
-    let setup_script = match shell_basename {
+    let setup_stub = match shell_basename {
         Some("bash") => {
-            let script = BASH_INIT;
+            /* This *should* look like the zsh function, but bash 3.2 (MacOS default shell)
```
Comment (matchai) [inline src/init.rs]: 👍

File: src/init.rs [hunk 2]
```diff
@@ -1,26 +1,39 @@
...
             Some(script)
         }
         Some("zsh") => {
-            let script = ZSH_INIT;
+            let script = "source <(starship init zsh --print-full-init)";
             Some(script)
         }
         Some("fish") => {
-            let script = FISH_INIT;
+            // Fish does process substitution with pipes and psub instead of bash syntax
```
Comment (matchai) [inline src/init.rs]: Huh... Neat! 👍

File: src/init.rs [hunk 3]
```diff
@@ -46,142 +59,179 @@ pub fn init(shell_name: &str) {
...
-   Finally, to work around existing DEBUG traps in the absence of a preexec-like,
-   we parse out the name of the old DEBUG hook, then make a new function which
-   calls both that function and our starship hooks. We don't do this for
-   PROMPT_COMMAND because that would probably result in two prompts.
-
-   We need to quote the output of `$(jobs -p | wc -l)` since MacOS `wc` leaves
-   giant spaces in front of the number (e.g. "    3"), which messes up the
-   word-splitting. Instead, quote the whole thing, then let Rust do the whitespace
-   trimming within the jobs module
+/* GENERAL INIT SCRIPT NOTES
```
Comment (matchai) [inline src/init.rs]: 👍

**Pull request merged**